### PR TITLE
refactor reloader.bash for better error handling

### PR DESCRIPTION
**Pull request by cornfeedhobo**
## Description

This PR seeks to:

- Apply linting for shfmt and shellcheck
- Ensure all tests use double bracket expansion (`[[]]`)
- Modify the loader loop to handle any scripts that may lead to an `exit`
    - This is done using recursion
    - Protects users from having a shell that unexpectedly exits before presenting a prompt
- **This approach has the side-effect of allowing plugin developers to not think about explicit `return`s**

_Devil's advocate:  It's possible that the complexity here isn't worth it. I think an alternate route could be to add an EXIT trap to `bash_it.sh` to short-circuit if a subshell leads to an `exit`, which would allow us to assume all plugins will only result in implicit or explicit `return`s, and reduce a lot of this complexity. I have not tested this idea though..._

## Motivation and Context

@NoahGorny asked about logging success conditions, and I said I'd rather centralize that. This addresses that, and more.

## How Has This Been Tested?

Locally

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] If my change requires a change to the documentation, I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [ ] If I have added a new file, I also added it to ``clean_files.txt`` and formatted it using ``lint_clean_files.sh``.
- [ ] I have added tests to cover my changes, and all the new and existing tests pass.

### CODE REVIEW

**Comment by cornfeedhobo** [comment]
I have an improvement that I want to investigate before merging. Hold off.

File: scripts/reloader.bash [hunk 1]
```diff
@@ -1,54 +1,81 @@
...
+		if [[ $retval -gt 0 ]]; then
+			_log_error 'Loading component failed...'
+			if [[ -n "$output" ]]; then
+				while read -r line; do _log_error "$line"; done <<< "$output"
+			fi
+		else
+			_log_debug 'Loading component successful!'
+			source "$current"
+			if [[ -n "$output" ]]; then
+				while read -r line; do _log_debug "$line"; done <<< "$output"
```
Comment (NoahGorny) [inline scripts/reloader.bash:41]: this also causes log messages in the plugin to be displayed twice (once regularly and once by this line)
we currently print logs to stdout, so maybe only catch output of stderr to display

**Review by NoahGorny — CHANGES_REQUESTED** [review]
I really like this change @cornfeedhobo, I have two comments though that needs to be addressed IMO

**Comment by NoahGorny** [comment]
ping @cornfeedhobo

**Comment by cornfeedhobo** [comment]
@NoahGorny I've had some family stuff came up. I'll get back to everything as soon as possible,

**Comment by NoahGorny** [comment]
> @NoahGorny I've had some family stuff came up. I'll get back to everything as soon as possible,

No worries, no need to feel rushed :smile:
I hope you are doing okay, and have a great day!

**Comment by cornfeedhobo** [comment]
@NoahGorny sorry. I need more time. I have other priorities at the moment.

**Comment by NoahGorny** [comment]
> @NoahGorny sorry. I need more time. I have other priorities at the moment.

I think you misunderstood me :cry: 
I meant that you do not need to feel rushed to do anything here, I totally appreciate the time you contribute to Bash-it, but thats your own time. If you need time for yourself, thats completely fine, I just hope you and your family situation will get better!

**Comment by cornfeedhobo** [comment]
@NoahGorny No problem at all. I emailed you to explain :)

File: scripts/reloader.bash (hunk 1 above)
[the remark this answers is not in the document]
Comment (cornfeedhobo) [reply scripts/reloader.bash:41]: @NoahGorny I tried re-writing this a few different ways and they are all verbose and less readable. I think this is the most readable format.

**Comment by cornfeedhobo** [comment]
I'm going to add some tests for this. Hold on to your butts

**Comment by cornfeedhobo** [comment]
My testing was going smooth until I added a canary to test for a silent failure. I'm going to take a different route. This might take a while. don't wait up.

**Comment by cornfeedhobo** [comment]
@NoahGorny Well, for now I've pushed the tests I'm working against. You let me know if these cases make sense to you.

File: scripts/reloader.bash (hunk 1 above)
[the remark this answers is not in the document]
Comment (cornfeedhobo) [reply scripts/reloader.bash]: this was actually intentional, but it appears that my double subshell method does not allow me to properly catch stdout and send it to the log. instead, everything still prints to the shell.

**Comment by NoahGorny** [comment]
Hi @cornfeedhobo , nice changes. I think that the test plugins you added should only be present when running the tests, maybe we should store them inside the `test` directory and copy them to the plugin directory on setup?

**Comment by cornfeedhobo** [comment]
@NoahGorny I think you might have looked at this on your phone or maybe missed that these plugins are in `test/fixtures/`

**Comment by NoahGorny** [comment]
Whoops, my bad. I will take a look now

**Comment by NoahGorny** [comment]
Hey @cornfeedhobo, I have made some small changes in order for everything to work, lemme know what you think

**Comment by cornfeedhobo** [comment]
@NoahGorny Sounds good. Let's leave this PR open for discussion before making any moves. I'm not in a rush and I'm not sure how I feel about the approach. I'm happy to protect users from exits, but I wish it wasn't at the cost of double loading, which I could see becoming expensive for plugins that do things like query brew, which has a very long start time.

**Comment by cornfeedhobo** [comment]
@NoahGorny oh, if you were trying to make the tests happy, they worked before I pushed and went to bed. Since these don't have pressure to make progress quickly, please don't feel the need to make edits or push it along. I'll get to the tests at my next free time. This is nowhere near the final state I'd like it in anyways. Thanks!

**Comment by cornfeedhobo** [comment]
@NoahGorny were the tests failing before you made your changes? Is that why you made edits?

**Comment by NoahGorny** [comment]
> @NoahGorny were the tests failing before you made your changes? Is that why you made edits?

Oh, my bad, I had weird "bash: trap: ./enabled/350---pip.completion.bash: invalid signal specification" lines whenever I reloaded bash-it, so I changed the trap line to use single quotes, this prevented the lines, but fails the test. I should have noticed that....sorry

Do you know why the test fails when using only single quotes in the trap line?

**Comment by cornfeedhobo** [comment]
The `scripts` variable would have to be a global for the trap to expand at the time of signalling, so I do the expansion up-front.

Looks like you got an error when manually testing?

**Comment by NoahGorny** [comment]
f

> The `scripts` variable would have to be a global for the trap to expand at the time of signalling, so I do the expansion up-front.
> 
> Looks like you got an error when manually testing?

I got this error every time I reloaded bash-it, in general usage.. I can add a pic if you want

**Comment by cornfeedhobo** [comment]
@NoahGorny I see it. I'll keep chipping away, but I'm not really in a rush. I'm probably going to pause on this and refocus on the language plugins, which is what I hear about the most with my users.

**Comment by gaelicWizard** [comment]
$0.02: I'm very much against subshells. They reduce performance, increase resource usage, and cause weird issues like making it difficult to redirect stdout. If we're running scripts from outside this project, then it entirely makes sense that they should be run independently rather than sourced, but if we're importing things from outside that are intended to improve our interactive shell experience (as is the purpose of this project) then we ought to demand that those outside scripts are reliable and predictable. 

I've actually just (barely) started on a PR to try to improve use of bash's own features to reduce calling out to external tools and subshells, but now that I've found this one I may need to rebase here and see how things go.

Alsö, I would heavily object to *any* sourced script *ever* calling `exit`. I'd come out mega-grumpy on their bug tracker.

**Comment by cornfeedhobo** [comment]
@gaelicWizard I gotta be honest, I haven't touched this in six months, and I'm completely confused by your comment and it's context.

Only your last paragraph made sense to me, and I can only say that I have no control over what a user does. My use case is I have users at a company, and they source a script with `exit` in it. They have zero control over this. I would much rather a loader that never leaves the shell unusable. Most users will not know how to debug once they find themselves with a shell that exits on load.

**Comment by gaelicWizard** [comment]
> My use case is I have users at a company, and they source a script with `exit` in it. They have zero control over this. I would much rather a loader that never leaves the shell unusable. Most users will not know how to debug once they find themselves with a shell that exits on load.

This makes so much more sense now. 

Alsö, literally just yesterday I found a cursed technique to abuse shell commands, so I may offer a simple (and fun!) PR that might address your use case. 😃

**Comment by gaelicWizard** [comment]
@cornfeedhobo 
this sounds mad silly...but...try this? It won't work if the `exit` is inside a function...
```
alias exit=return
source /the/bad/script
unalias exit
```
Does that work?

**Comment by cornfeedhobo** [comment]
It's crazy, but I'm willing to explore the idea. Thanks for sharing!!

Luckily, this isn't a rushed item. I was able to work with each user to identify and fix their problems, but it's what made me consider a generalized loader to handle this.

**Comment by NoahGorny** [comment]
interesting- I am waiting for a followup from this :smile:

File: scripts/reloader.bash [hunk 2]
```diff
@@ -1,54 +1,84 @@
-#!/bin/bash
+# shellcheck shell=bash
+
 BASH_IT_LOG_PREFIX="core: reloader: "
-pushd "${BASH_IT}" >/dev/null || exit 1
+pushd "${BASH_IT}" > /dev/null || exit 1
+
+_bash-it-load-sources() {
```
Comment (gaelicWizard) [inline scripts/reloader.bash:6]: Making this a function ruins all explicit `declare`s, `typeset`s, &c. Bash makes any variables set this way as local to the function, which I suspect is why `reloader.bash` is a script file in the first place.

**Comment by davidpfarrell** [comment]
Sorry I missed the fun here but wanted to share my ultimate exit vs return hack:

_exit-or-return.sh_
```
#!/usr/bin/env bash

(return 0 2> /dev/null) && return ${1:-0} || exit ${1:-0}
```

_usage_
```
$ /tmp/exit-or-return.sh ; echo $?
0

$ /tmp/exit-or-return.sh 7 ; echo $?
7

$ source /tmp/exit-or-return.sh ; echo $?
0

$ source /tmp/exit-or-return.sh 127 ; echo $?
127
```

Cheers !

**Comment by gaelicWizard** [comment]
Here's what I use to determine if the script is being sourced or executed:
```
stuff() { :; }

if [[ "${BASH_SOURCE[0]}" != "$0" ]]
then # We are being sourced.
: # do anything necessary for external usage
else # we are being executed 
: # run main function with "$@" and exit
fi
```

I like it because I can have one file with both ends of a client/server setup (e.g., `screen`) functions and whatnot.

**Pull request closed without merging**